### ISSUE TITLE
BUG: v1.13 introduced undocumented change in `sparse.csr_matrix` creation of matrices containing trailing column of all zeros

### ISSUE THREAD

**Issue by RUrlus** [scipy/scipy#20428]
### Describe your issue.

Release v1.13 caused a failure in our scheduled test-suite because it introduced a (in our case breaking) change in the instantiation of `csr_matrix` from a tuple of `data, indices, indptr` without specifying the shape.

I suspect this is an edge case that we happened to hit, but I wanted to report it anyway in case others encounter it or it is of interest.
There is a good chance that the previous behaviour was wrong/undesirable as the current behaviour is consistent the across my attempts to create a MRE.
It does, however, appear to be an undocumented change as I didn't read anything in the release node that could explain the change.

I've not been able to replicate outside our (very) specific test-case but the issue was encountered consistently across multiple platforms and Python versions, see [CI with Scipy 1.13.0](https://github.com/ing-bank/sparse_dot_topn/actions/runs/8607751349/job/23588839411) which passed last week with [Scipy 1.12](https://github.com/ing-bank/sparse_dot_topn/actions/runs/8514885669/job/23321396844)

The issue (appears) to be caused by a change to shape determination of the `csr_matrices` when the last column of the matrix contains all zeros.
In our case we test the reconstruction of a sparse matrix after splitting up the dot product and taking the top-n values.
One of the sub-matrices contains an all-zero column which results in differently shaped matrices depending on the Scipy version: 
* v1.12 the shape of the matrix is `200x600`
* v1.13 the shape of the matrix is `200x599`

Note that both matrices have the same amount of non-zero entries. Setting the desired shape during the creation of the matrix resolves the problem, e.g. `csr_matrix((data, indices, indptr), shape=(200, 600))`
Note that MRE requires our package `sparse-dot-topn` (available on PyPi).
Creating a MRE outside our test case that exhibits the same behaviour appears to be very tricky.

Feel free to close the issue if you feel that the current behaviour is the correct behaviour.

### Reproducing Code Example

```python
# NOTE REQUIRES sparse-dot-topn v1.1.0 to replicate
# pip install sparse_dot_topn==1.1.0

import numpy as np
from scipy import sparse
from sparse_dot_topn import sp_matmul_topn, zip_sp_matmul_topn

rng = np.random.Generator(np.random.PCG64DXSM())
# replicate state of RNG during the test as in our test-suite
rng.bit_generator.state = {
    'bit_generator': 'PCG64DXSM',
    'state': {
        'state': 95283714116121921517540898303641654026,
        'inc': 327998138400881660830365005857473961283
    },
    'has_uint32': 0,
    'uinteger': 1607172753
}

A = sparse.random(1000, 2000, density=0.1, format="csr", dtype=np.float32, random_state=rng)
B = sparse.random(600, 2000, density=0.1, format="csr", dtype=np.float32, random_state=rng)

# zipped C-matrix
As = [A[i * 200 : (i + 1) * 200] for i in range(5)]
Bs = [B[:100], B[100:300], B[300:]]

Cs = [[sp_matmul_topn(Aj, Bi.T, top_n=10, threshold=0.01, sort=True) for Bi in Bs] for Aj in As]
C_zips = [zip_sp_matmul_topn(top_n=10, C_mats=Cis) for Cis in Cs]

# test case fails on vstack call
# C_stack = sparse.vstack(C_zips, dtype=dtype)

# below is true with v1.12 and false with 1.13
assert C_zips[-1].shape == (200, 600)
```

### Error message

```shell
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../../.pyenv/versions/3.11.3/envs/SDTOPN/lib/python3.11/site-packages/scipy/sparse/_construct.py:781: in vstack
    return _block([[b] for b in blocks], format, dtype, return_spmatrix=True)
../../../../.pyenv/versions/3.11.3/envs/SDTOPN/lib/python3.11/site-packages/scipy/sparse/_construct.py:912: in _block
    A = _compressed_sparse_stack(blocks[:, 0], 0, return_spmatrix)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

blocks = array([<200x599 sparse matrix of type '<class 'numpy.float32'>'
        with 2000 stored elements in Compressed Sparse...type '<class 'numpy.float32'>'
        with 2000 stored elements in Compressed Sparse Row format>],
      dtype=object)
axis = 0, return_spmatrix = True

    def _compressed_sparse_stack(blocks, axis, return_spmatrix):
        """
        Stacking fast path for CSR/CSC matrices or arrays
        (i) vstack for CSR, (ii) hstack for CSC.
        """
        other_axis = 1 if axis == 0 else 0
        data = np.concatenate([b.data for b in blocks])
        constant_dim = blocks[0].shape[other_axis]
        idx_dtype = get_index_dtype(arrays=[b.indptr for b in blocks],
                                    maxval=max(data.size, constant_dim))
        indices = np.empty(data.size, dtype=idx_dtype)
        indptr = np.empty(sum(b.shape[axis] for b in blocks) + 1, dtype=idx_dtype)
        last_indptr = idx_dtype(0)
        sum_dim = 0
        sum_indices = 0
        for b in blocks:
            if b.shape[other_axis] != constant_dim:
>               raise ValueError(f'incompatible dimensions for axis {other_axis}')
>       _assert_array_equal(C_stack.indices, C_ref.indices)
```

### SciPy/NumPy/Python version and system information

```shell
1.13.0 1.26.3 sys.version_info(major=3, minor=11, micro=3, releaselevel='final', serial=0)

Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/arm64-builds/include
    lib directory: /opt/arm64-builds/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=64
    pc file directory: /opt/arm64-builds/lib/pkgconfig
    version: 0.3.26.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/arm64-builds/include
    lib directory: /opt/arm64-builds/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=64
    pc file directory: /opt/arm64-builds/lib/pkgconfig
    version: 0.3.26.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 15.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 15.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld64
    name: gcc
    version: 13.2.0
  pythran:
    include directory: ../../../../../../private/var/folders/l_/53d72k7j4zl9kj8fzncsd8hw0000gn/T/pip-build-env-xo17zr8z/overlay/lib/python3.11/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
  cross-compiled: false
  host:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
Python Information:
  path: /private/var/folders/l_/53d72k7j4zl9kj8fzncsd8hw0000gn/T/cibw-run-95i9ncig/cp311-macosx_arm64/build/venv/bin/python
  version: '3.11'
```

**Comment by j-bowhay** [comment]
cc @dschult

**Comment by perimosocordiae** [comment]
Apologies for the delayed response. Here's what I tried first, on both 1.13 and 1.12:

```python
import numpy as np
from scipy.sparse import csr_matrix

a = np.eye(2, 4)  # right two columns are all zero
b = csr_matrix(a)
assert b.shape == (2, 4)
c = csr_matrix((b.data, b.indices, b.indptr))
assert c.shape == (2, 2)
```

In both cases, the `c` matrix has only two columns (as expected).

I also tried running your reproducer in both versions, and observed the difference in shape. The contents of the sparse matrices are different, though, so I don't think the discrepancy is in how scipy is initializing the CSR matrix, but rather in some other factor. My suspicion is that the difference is coming from the `sparse.random()` calls, as that module has seen some changes recently that may affect the order of RNG accesses.

That said, this doesn't look like a scipy issue: if no shape is provided, we assume that the number of columns matches the largest explicit column index. Your fix that specifies the shape explicitly is the correct way to handle the possibility of a right-most column being empty.
